grafana-stack/deploy

```diff
@@ -0,0 +1,230 @@
+          gcloud compute disks create "${name}" --size "${size}" &&
+  apply_config "${vclaim}" &&
+  apply_config "${volume}"
+}
+
+prepare_nginx() {
+  nginx_tag=$(date +%s)
+  nginx_repo=${REGISTRY}/nginx
+}
+
+prepare_grafana() {
+  grafana_tag=$(date +%s)
+  grafana_repo=${REGISTRY}/grafana
+}
+
+prepare_influxdb() {
+  influxdb_tag=$(date +%s)
+  influxdb_repo=${REGISTRY}/influxdb
+}
+
+get_public_ip() {
+  service=$1
+
+  kubectl --kubeconfig="${KUBECONFIG}" get svc "${service}" -o=jsonpath="{.status.loadBalancer.ingress[0].ip}"
+}
+
+wait_nginx_service() {
+  for i in $(seq 6); do
+    ip=$(get_public_ip nginx-service) &&
+    if test ! -z ${ip}; then
+      print_line "Nginx service is up:" ${ip}
+      return
+    fi
+    sleep 5
+  done
+  false
+}
+
+create_grafana_configs() {
+  grafana_admin_password=$1
+
+  sed -e "s/admin-password-placeholder/${grafana_admin_password}/" grafana.template.ini >grafana.ini
+}
+
+prepare_nginx_deployment() {
+  nginx_image=$1
+  sed -e "s#nginx-image-placeholder#${nginx_image}#" nginx.template.yaml >nginx.yaml
+}
+
+prepare_grafana_deployment() {
+  grafana_image=$1
+  sed -e "s#grafana-image-placeholder#${grafana_image}#" grafana.template.yaml >grafana.yaml
+}
+
+prepare_influxdb_deployment() {
+  influxdb_image=$1
+  root_password=$2
+
+  sed \
+    -e "s#influxdb-image-placeholder#${influxdb_image}#" \
+    -e "s/influxdb-root-password-placeholder/${root_password}/" \
+    influxdb.template.yaml >influxdb.yaml
+}
+
+create_grafana_datasource() {
+  server_hostname=${1:-$(get_public_ip nginx-service)} &&
+  grafana_admin_password=$2 &&
+
+  curl -s --fail "http://admin:${grafana_admin_password}@${server_hostname}/api/datasources/name/github" ||
+  cat datasource.json | env - server_hostname=${server_hostname} envsubst | curl -X POST --fail -H 'Content-Type: application/json' -H 'Accept: application/json' "http://admin:${grafana_admin_password}@${server_hostname}/api/datasources" --data-binary @-
+}
+
+create_influxdb_user() {
+  server_hostname=${1:-$(get_public_ip nginx-service)} &&
+  influxdb_root_password=$2 &&
+
+  cat create_user | env - server_hostname=${server_hostname} influx_db_root_password=${influxdb_root_password} envsubst | sh
+}
+
+has $grafana && {
+  cd grafana &&
+  run_command "Preparing grafana environment" prepare_grafana
+  run_command "Update grafana configurations" create_grafana_configs "${GRAFANA_ADMIN_PASSWORD}"
+  run_command "Create/Update grafana PD" create_pd grafana-database grafana_vclaim.yaml grafana_volume.yaml 2GiB
+  run_command "Create/update grafana container" create_container "${grafana_repo}:${grafana_tag}" Dockerfile
+  run_command "Push grafana container" push_container "${grafana_repo}:${grafana_tag}"
+  run_command "Prepare deployment configuration" prepare_grafana_deployment "${grafana_repo}:${grafana_tag}"
+  run_command "Deploy grafana" apply_config grafana.yaml
+  run_command "Create grafana service" apply_config grafana_service.yaml
+  cd ..
```

Comment:
OK I know. Without errexit, a failing run_command would do `exit` and exit the subshell but not the parent shell.